### ISSUE TITLE
Handle ratelimiting consistently

### ISSUE THREAD

**Issue by eritbh** [toolbox-team/reddit-moderator-toolbox#248]
The only real ratelimiting code toolbox actually has is in [TBCore.forEachChunkedRateLimit](), which from what I can tell basically says "here's a function that'll make x API calls, run it y times as fast as you can but don't let the `Ratelimit-Remaining` header dip below 20 and if it does stop everything for six minutes". It was added [back in 2015](https://github.com/toolbox-team/reddit-moderator-toolbox/commit/92298a628feb2aca4bf76c9435a7d1f87e28e99a#diff-c9da5bc7882bb894de2a63a2aaede53aR1027-R1098) and hasn't significantly changed since.

This system provides only very weak protection, relies on tons of assumptions, is written in a very messy and hard-to-understand way, provides no protection for API calls that are made outside this function, and [may just not work as intended at all](https://discordapp.com/channels/535490452066009090/535490452066009093/685253706178101312) when a large amount of requests are required (e.g. usernote pruning by activity).

Anyway enough ranting about that. Ratelimiting should be handled in the background page, by the request handler, by actually reading the ratelimit headers for every request and queueing requests to be sent if more stack up than can be handled in the current ratelimit bucket. Care should be taken to ensure that `ratelimit-reset` headers are interpreted properly so we don't end up waiting longer for ratelimits than we have to.

This system will protect *all* requests from ratelimit-related failure and will obviate `TBCore.forEachChunkedRateLimit`. It can then be removed and replaced by simple `for` loops, and we can make sure that appropriate loading indicators are displayed for all tasks that may take a long time to complete due to frequent limiting.

**Comment by creesch** [comment]
Not going to deny it is messy, though I am not sure where you are getting the 6 minutes from as it should just take the time reddit gives for cooldown. However that very well could be 6 minutes .


Other than that, I am all for a system of handling ratelimiting consistently. Is it maybe an idea to do this as part of #45?

**Comment by Geo1088** [comment]
> though I am not sure where you are getting the 6 minutes from

I'm not sure where it's coming from either, lol. That was an observation I made when testing it yesterday - I posted the logs in discord, `ratelimitReset` was reporting 8 or 9 seconds but the timer told me it was gonna be multiple minutes before anything resumed. The logic behind determining the number of seconds left is very confusing to me and I haven't been able to identify what's actually happening. At this point though, I'm not gonna bother trying to understand the current code, I'll just plan to rip it out and re-implement rate limiting entirely.

I think this and #45 compliment each other well, though I'll have to address the removal of the existing `forEachChunkedRateLimit` function separately. I'll probably do two PRs - one for the basics of #45 to make sure that our code handles `Response` objects instead of `jqXHR` objects, and another to reimplement ratelimiting on the background page and remove uses of `forEachChunkedRateLimit`.

**Comment by creesch** [comment]
Sounds good to me. I assume we are going for a reactive sort of rate limit handling (as opposed to the semi-proactive one we have now)? Basically something like 

1. do request.
2. If response is "nope ratelimited" put the current request in a queue for handling based the timeout given by reddit.
3. put all subsequent requests in the queue until the ratelimit refresh time is done. 

Or maybe put all requests into a handle queue and pause handling as needed?  Something along the lines of the very much pseudo code below (interval just used to demonstrate the concept)

```javascript
    let requestBuffer = [];
    let rateLimitPauze = false;

    setInterval(async () =>{
        if(requestBuffer.length && !rateLimitPauze) {
            const request = requestBuffer.shift();

            const response = await fetch(request);

            if(response.status === 'ratelimit bullshit') {
                rateLimitPauze = true;
                requestBuffer.unshift(request);
                setTimeout(() => {
                    rateLimitPauze = false;
                }, response.headers.ratelimitTimeThingyHeader)
            }

        }

    }, 5);

    browser.runtime.onMessage.addListener(async request => {
        requestBuffer.push(request);
    });


```

If my assumption is correct, do we want to implement a flag where requests can bypass ratelimiting entirely? I am not entirely sure if we have them but I can imagine requests where it is not vital that they go through...

**Comment by Geo1088** [comment]
The idea is to maintain a list, and rather than reacting only to 429s, we read the ratelimit headers sent with the response of *every* request. If the ratelimit remaining is ever 0, then the request manager waits the reset period even before encountering a 429. If a 429 is encountered anyway, then the request gets retried after the reset period.

We shouldn't allow requests to bypass ratelimits - under this system, a request set to bypass would never actually work anyway, since we always use the entire ratelimit before ever pausing the queue.

**Comment by creesch** [comment]
Right yeah, getting a 0 limit back would also be a valid moment to pause the thing :) Concept remains the same right?

**Comment by k3n** [comment]
@SpyTec follow-up to the convo started in Discord re: feedback on multiple actions.

Brainstorming here, so beware this isn't a fully-formed idea, but I was thinking of the benefits that this unified rate-limiting feature would bring us via providing a central location to inject additional behaviors. 

For instance, I'd presume whatever form this code takes will have some sort of queue, and would know the status of each item in that queue. Given such, we could expose what we know in a meaningful manner within some place like the TB context menu or menu bar. 

Maybe a badge or some other UI affordance with the current count of outstanding requests, which could additionally be decorated to communicate that the rate limit has been reached and requests are queued, like the number (or background) turns red, the badge bounces, or something to that effect.

Maybe hovering this badge would give more info, e.g.:

    6/10 requests processed (11s until next request)

Taking this wacky idea further, clicking the badge -- and perhaps, auto-opening when a mass-request action is initiated, e.g. multi-selected modqueue items -- could open a pop-up of some sort that'd list each request and their status.

Indulge me...

Key:
    ✔ complete
    ❌ failed (for completeness, may not need)
    ◌ (animated) in progress
    ○ todo

     _______________________________________________________________
    |  TB Request Manager                                           |
    |---------------------------------------------------------------|
    |     Progress (2/6 complete, 2 in progress):                   |
    |     __________________________________________                |
    |    |=============                             |               |
    |     ------------------------------------------                |
    |                                                               |
    |  ✔  Remove: "Title of the 1st post in the queue" by u/one     |
    |  ✔  Remove: "Title of the 2nd post" by u/two                  |
    |  ◌  Spam: "Title of the 3rd post" by u/three                  |
    |  ◌  Spam: "Title of 4th post" by u/four                       |
    |  ○  Remove: "Title of 5th post" by u/five                     |
    |  ○  Spam: "Title of 6th post" by u/six                        |
    |                         ____________     ____________         |
    |                        | Cancel All |   |     OK     |        |
    |                         ------------     ------------         |
     ---------------------------------------------------------------

The proposed UI implementations are just for conversation fodder, the main idea that I think we could leverage here is the unified request interface for rate-limiting. We have a lot of possibilities!

**Comment by creesch** [comment]
Sounds pretty neat though would require a lot more work than is planned now as it also then needs to not only log the request but also do so in a way that we can show in human readable format to our users. Because reddit api endpoints don't always make sense. 

I could see some use for it in a debug setting. 

Btw, that is some neat ascii art!

**Comment by k3n** [comment]
I definitely thought of it as a debug item originally, but then I had the question about feedback on multi-actions and it struck me that it could be used to give feedback to the user as well. 

And sorry @Geo1088 for veering off-topic, I just really like your idea and consider follow-on improvements that would be possible as a _yes_ vote for this feature :)

**Comment by Geo1088** [comment]
I really like that idea. Actually implementing a UI like that is probably out of scope of this PR, but exposing request progress is definitely doable.

When creesch and I discussed this, we figured it would be enough to just have the request handler emit an event if the queue got delayed for a limit, but for something more complicated like you're proposing we'd need a more complex system that would assign each pending request some sort of ID and then emit events when each request's status changes. It will require a bit more architecting, but definitely possible if we want to!
